securecookie.go

```diff
@@ -205,27 +206,32 @@ func checkPath(path string) error {
 // See https://tools.ietf.org/html/rfc1034#section-3.5 and
 // https://tools.ietf.org/html/rfc1123#section-2.
 func checkDomain(name string) error {
-	if len(name) > 255 {
+	switch {
+	case len(name) == 0:
+		return nil // an empty domain name will result in a cookie without a domain restriction
+	case len(name) > 255:
 		return fmt.Errorf("cookie domain: name length is %d, can't exceed 255", len(name))
-	}
-	for i := 0; i < len(name); i++ {
-		c := name[i]
+	case name[0] == '.':
+		return errors.New("cookie domain: starts with '.'")
+	case isDigit(name[0]):
+		return errors.New("cookie domain: starts with digit")
+	case name[len(name)-1] == '.':
+		return errors.New("cookie domain: ends with '.'")
+	}
+	for i, c := range name {
+		if utf8.RuneLen(c) != 1 {
+			return fmt.Errorf("cookie domain: invalid character '%c' at offset %d", c, i)
+		}
 		if c == '.' {
```

Comment:
I would use a byte variable after this for clarity:
ˋb := byte(c)ˋ

Reply:
Or use a rune everywhere which would be simpler. I don't think it would harm performance.